grakn/service/Session/util/Communicator.py

```diff
@@ -0,0 +1,143 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import six
+from collections import deque
+from six.moves import queue
+from grakn.exception.GraknError import GraknError
+
+
+class SingleResolver:
+    def __init__(self, communicator, request):
+        self._request = request
+        self._communicator = communicator
+        self._buffered_response = None
+        communicator._send_with_resolver(self, request)
+
+    def _is_last_response(self, response):
+        return True
+
+    def _buffer_response(self, response):
+        self._buffered_response = response
+
+    def get(self):
+        response = self._buffered_response
+        if response:
+            return response
+        else:
+            response = self._communicator._block_for_next(self)
+            self._buffered_response = response
+            return response
+
+
+class IterationResolver(six.Iterator):
+    def __init__(self, communicator, request, is_last_response):
+        self._request = request
+        self._is_last_response = is_last_response
+        self._communicator = communicator
+        self._ended = False
+        self._response_buffer = deque()
+        communicator._send_with_resolver(self, request)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            return self._response_buffer.popleft()
+        except IndexError:
+            if self._ended:
+                raise StopIteration()
+            return self._communicator._block_for_next(self)
+
+    def _buffer_response(self, response):
+        self._response_buffer.append(response)
+
+    def _end(self):
+        self._ended = False
+
+
+class Communicator(six.Iterator):
+    def __init__(self, grpc_stream_constructor):
+        self._request_queue = queue.Queue()
+        self._resolver_queue = deque()
+        self._response_iterator = grpc_stream_constructor(self)
+        self._closed = False
+
+    def __iter__(self):
+        return self
+
+    # Used by the GRPC stream to iterate requests as they arrive
+    def __next__(self):
+        request = self._request_queue.get(block=True)
+        if request is None:
+            raise StopIteration()
+        return request
+
+    def error_if_closed(self):
+        if self._closed:
+            raise GraknError("This connection is closed")
+
+    # Put a request for GRPC to consume
+    def _send_with_resolver(self, resolver, request):
+        self._resolver_queue.append(resolver)
+        self._request_queue.put(request)
+
+    def _block_for_next(self, resolver):
```

Comment:
since i think the meat of the demultiplexing of the stream of responses happens here, would be good to write a docstring with a small explanation.

side note, i think way you've got the "parallel" lists of request/resolver lists to keep track of which resolver to give answers, taking advantage of the gRPC ordering guarantees is neat!

Reply:
I added it at the top of the file, I think that *should* suffice, what do you think?

Reply:
yep